pkg/controller/backup_sync_controller.go

```diff
@@ -21,378 +21,408 @@ import (
+			velerov1api.StorageLocationLabel: label.GetValidName(locationName),
+		}).AsSelector(),
+	}
+	err := b.client.List(ctx, &backupList, &listOption)
 	if err != nil {
 		log.WithError(errors.WithStack(err)).Error("Error listing backups from cluster")
 		return
 	}
-	if len(backups) == 0 {
+
+	if len(backupList.Items) == 0 {
 		return
 	}
 
-	for _, backup := range backups {
+	for _, backup := range backupList.Items {
 		log = log.WithField("backup", backup.Name)
 		if backup.Status.Phase != velerov1api.BackupPhaseCompleted || backupStoreBackups.Has(backup.Name) {
 			continue
 		}
-		if err := c.backupClient.Backups(backup.Namespace).Delete(context.TODO(), backup.Name, metav1.DeleteOptions{}); err != nil {
+
+		if err := b.client.Delete(ctx, &backup, &client.DeleteOptions{}); err != nil {
 			log.WithError(errors.WithStack(err)).Error("Error deleting orphaned backup from cluster")
 		} else {
 			log.Debug("Deleted orphaned backup from cluster")
-			c.deleteCSISnapshotsByBackup(backup.Name, log)
+			b.deleteCSISnapshotsByBackup(ctx, backup.Name, log)
 		}
 	}
 }
 
-func (c *backupSyncController) deleteCSISnapshotsByBackup(backupName string, log logrus.FieldLogger) {
+func (b *backupSyncReconciler) deleteCSISnapshotsByBackup(ctx context.Context, backupName string, log logrus.FieldLogger) {
 	if !features.IsEnabled(velerov1api.CSIFeatureFlag) {
 		return
 	}
 	m := client.MatchingLabels{velerov1api.BackupNameLabel: label.GetValidName(backupName)}
-	if vsList, err := c.csiVSLister.List(label.NewSelectorForBackup(label.GetValidName(backupName))); err != nil {
+	var vsList snapshotv1api.VolumeSnapshotList
+	listOptions := &client.ListOptions{
+		LabelSelector: label.NewSelectorForBackup(label.GetValidName(backupName)),
+	}
+	if err := b.client.List(ctx, &vsList, listOptions); err != nil {
 		log.WithError(err).Warnf("Failed to list volumesnapshots for backup: %s, the deletion will be skipped", backupName)
 	} else {
-		for _, vs := range vsList {
+		for _, vs := range vsList.Items {
 			name := kube.NamespaceAndName(vs.GetObjectMeta())
 			log.Debugf("Deleting volumesnapshot %s", name)
-			if err := c.kbClient.Delete(context.TODO(), vs); err != nil {
+			if err := b.client.Delete(context.TODO(), &vs); err != nil {
 				log.WithError(err).Warnf("Failed to delete volumesnapshot %s", name)
 			}
 		}
 	}
 	vsc := &snapshotv1api.VolumeSnapshotContent{}
 	log.Debugf("Deleting volumesnapshotcontents for backup: %s", backupName)
-	if err := c.kbClient.DeleteAllOf(context.TODO(), vsc, m); err != nil {
+	if err := b.client.DeleteAllOf(context.TODO(), vsc, m); err != nil {
 		log.WithError(err).Warnf("Failed to delete volumesnapshotcontents for backup: %s", backupName)
 	}
 }
+
+// backupSyncSourceOrderFunc returns a new slice with the default backup location first (if it exists),
+// followed by the rest of the locations in no particular order.
+func backupSyncSourceOrderFunc(objList client.ObjectList) client.ObjectList {
```

Comment:
The input `objList` should can be converted to `*velerov1api.BackupStorageLocationList{}` directly, so that we can iterator its items and find the default BSL, and replace the default BSL element with items[0]?

Reply:
Modified.